ai_ref_kits/agentic_llm_rag/main.py

```diff
@@ -0,0 +1,429 @@
+
+            gr.Markdown("------------------------------")            
+
+        demo.launch()
+
+    run()
+
+
+def main(chat_model: str, embedding_model: str, rag_pdf: str, device: str):
+    # Load models and embedding based on parsed arguments
+    llm, embedding = setup_models(chat_model, embedding_model, device)
+
+    Settings.embed_model = embedding
+    Settings.llm = llm
+
+    # Set up tools
+    paint_cost_calculator, add_to_cart_tool, get_cart_items_tool, clear_cart_tool, paint_gallons_calculator = setup_tools()
+    
+    text_example_en_path = Path(rag_pdf)
+    index = load_documents(text_example_en_path)
+    log.info(f"loading in {index}")
+ 
+    vector_tool = QueryEngineTool(
+        index.as_query_engine(streaming=True),
+        metadata=ToolMetadata(
+            name="vector_search",
+            description="""            
+            Use this tool for ANY question about paint products, recommendations, prices, or technical specifications.
+            
+            WHEN TO USE:
+            - User asks about paint types, brands, or products
+            - User needs price information before adding to cart
+            - User needs recommendations based on their project
+            - User has technical questions about painting
+            
+            EXAMPLES:
+            - "What paint is best for kitchen cabinets?"
+            - "How much does AwesomePainter Interior Acrylic Latex cost?"
+            - "What supplies do I need for painting my living room?"
+            """,
+        ),
+    )
+    
+    nest_asyncio.apply()
+ 
+    # Define agent and available tools
+    agent = ReActAgent.from_tools(
+        [paint_cost_calculator, add_to_cart_tool, get_cart_items_tool, clear_cart_tool, vector_tool, paint_gallons_calculator],
+        llm=llm,
+        max_iterations=5,  # Set a max_iterations value
+        handle_reasoning_failure_fn=custom_handle_reasoning_failure,
+        verbose=True,
+        react_chat_formatter=ReActChatFormatter.from_defaults(
+            observation_role=MessageRole.TOOL   
+        ),
+    ) 
+    react_system_prompt = PromptTemplate(react_system_header_str)
+    agent.update_prompts({"agent_worker:system_prompt": react_system_prompt})  
+    agent.reset()                     
+    run_app(agent)
+
+if __name__ == "__main__":
+    # Define the argument parser at the end
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--chat_model", type=str, default="model/qwen2-7B-INT4", help="Path to the chat model directory")
+    parser.add_argument("--embedding_model", type=str, default="model/bge-large-FP32", help="Path to the embedding model directory")
+    parser.add_argument("--rag_pdf", type=str, default="data/test_painting_llm_rag.pdf", help="Path to a RAG PDF file with additional knowledge the chatbot can rely on.")    
+    parser.add_argument("--device", type=str, default="GPU", help="Device for inferencing (CPU,GPU,GPU.1,NPU)")

```

Comment:
As this is a gradio demo, it should also have a "public" parameter. Please check Conversational AI Chatbot for more details